File: innerlibs1/Select2Data.vb
Imports System.Runtime.CompilerServices
Imports InnerLibs.LINQ

Namespace Select2

    Public Class Select2Data

        Sub New()

        End Sub

        Sub New(Options As IEnumerable(Of ISelect2Option))
            Me.Results = If(Options, {})
        End Sub

        Sub New(Groups As IEnumerable(Of Select2Group))
            Me.Results = If(Groups, {})
        End Sub

        Public Property Results As IEnumerable(Of ISelect2Result) = {}
        Public Property Pagination As New Pagination
    End Class

    Public Interface ISelect2Result
        Property Text As String

    End Interface

    Public Class Pagination
        Public Property More As Boolean = False
    End Class

    Public Class Select2Group
        Implements ISelect2Result

        Sub New(Text As String)
            Me.Text = Text
        End Sub

        Sub New(Text As String, Children As IEnumerable(Of ISelect2Option))
            Me.Text = Text
            Me.Children = If(Children, {})
        End Sub

        Sub New()

        End Sub

        Public Property Children As IEnumerable(Of ISelect2Option) = {}

        Public Property Text As String Implements ISelect2Result.Text

    End Class

    Public Interface ISelect2Option
        Inherits ISelect2Result

        Property ID As String
        Property Selected As Boolean

        Property Disabled As Boolean
    End Interface

    Public NotInheritable Class Select2Option
        Implements ISelect2Option

        Sub New()

        End Sub

        Sub New(Text As String, Value As String)
            Me.ID = Value.IfBlank(Text)
            Me.Text = Text
        End Sub

        Public Property ID As String Implements ISelect2Option.ID

        Property Selected As Boolean Implements ISelect2Option.Selected

        Property Disabled As Boolean Implements ISelect2Option.Disabled

        Public Property Text As String Implements ISelect2Result.Text

        Public Overrides Function ToString() As String
            Return $"<option value='{ID}' {Disabled.AsIf("disabled")} {Selected.AsIf("selected")}>{Text}</option>"
        End Function

    End Class

    Public Module Select2Extensions

        <Extension> Public Function CreateSelect2Data(Of OptionType As ISelect2Option, T)(List As IEnumerable(Of T), TextSelector As Func(Of T, String), IdSelector As Func(Of T, String), Optional OtherSelectors As Action(Of T, OptionType) = Nothing, Optional GroupSelector As Func(Of T, String) = Nothing) As Select2Data
            If GroupSelector IsNot Nothing Then
                Dim itens = List.GroupBy(GroupSelector).Select(Function(x) New Select2Group(x.Key, x.Select(Function(c) c.CreateSelect2Option(TextSelector, IdSelector, OtherSelectors))))
                Return New Select2Data(itens)
            Else
                Dim itens = List.Select(Function(c) c.CreateSelect2Option(TextSelector, IdSelector, OtherSelectors))
                Return New Select2Data(itens)
            End If
        End Function

        <Extension()> Function CreateSelect2Option(Of OptionType As ISelect2Option, T)(item As T, TextSelector As Func(Of T, String), IdSelector As Func(Of T, String), Optional OtherSelectors As Action(Of T, OptionType) = Nothing) As OptionType
            If GetType(T) Is GetType(OptionType) Then
                Return item.ChangeType(Of OptionType)
            Else
                IdSelector = If(IdSelector, TextSelector)
                TextSelector = If(TextSelector, IdSelector)
                Dim Optionitem = Activator.CreateInstance(Of OptionType)
                Optionitem.ID = IdSelector(item)
                Optionitem.Text = TextSelector(item)
                If OtherSelectors IsNot Nothing Then
                    OtherSelectors(item, Optionitem)
                End If
                Return Optionitem
            End If
        End Function

        <Extension> Public Function CreateSelect2Data(Of OptionsType As ISelect2Option, T1 As Class, T2)(Filter As PaginationFilter(Of T1, T2), TextSelector As Func(Of T2, String), IdSelector As Func(Of T2, String), Optional GroupSelector As Func(Of T2, String) = Nothing, Optional OtherSelectors As Action(Of T2, OptionsType) = Nothing) As Select2Data
            Dim d = Filter.GetPage().CreateSelect2Data(TextSelector, IdSelector, OtherSelectors, GroupSelector)
            d.Pagination.More = Filter.PageCount > 1 AndAlso Filter.IsLastPage = False
            Return d
        End Function

        <Extension> Public Function CreateSelect2Data(Of OptionsType As ISelect2Option, T1 As Class, T2)(Filter As PaginationFilter(Of T1, T2), TextSelector As Func(Of T2, String), IdSelector As Func(Of T2, String), OtherSelectors As Action(Of T2, OptionsType)) As Select2Data
            Return Filter.CreateSelect2Data(TextSelector, IdSelector, Nothing, OtherSelectors)
        End Function

        <Extension> Public Function CreateSelect2Data(Of OptionsType As ISelect2Option, T1 As Class)(Filter As PaginationFilter(Of T1, OptionsType)) As Select2Data
            Return Filter.CreateSelect2Data(Of OptionsType)(Function(x As OptionsType) x.Text, Function(x As OptionsType) x.ID)
        End Function

        <Extension> Public Function CreateSelect2Data(Of OptionsType As ISelect2Option, T1 As Class)(Filter As PaginationFilter(Of T1, OptionsType), GroupBySelector As Func(Of OptionsType, String)) As Select2Data
            Return Filter.CreateSelect2Data(Of OptionsType)(Function(x As OptionsType) x.Text, Function(x As OptionsType) x.ID, GroupBySelector)
        End Function

    End Module
End Namespace

File: InnerLibs/Select2Data.cs
using InnerLibs.LINQ;
using System;
using System.Collections.Generic;
using System.Linq;

namespace InnerLibs.Select2
{
    public interface ISelect2Option : ISelect2Result
    {
        bool Disabled { get; set; }
        string ID { get; set; }
        bool Selected { get; set; }
    }

    public interface ISelect2Result
    {
        string Text { get; set; }
    }

    public static class Select2Extensions
    {
        public static Select2Data CreateSelect2Data<OptionType, T>(this IEnumerable<T> List, Func<T, string> TextSelector, Func<T, string> IdSelector, Action<T, OptionType> OtherSelectors = null, Func<T, string> GroupSelector = null) where OptionType : ISelect2Option
        {
            if (GroupSelector != null)
            {
                var itens = List.GroupBy(GroupSelector).Select(x => new Select2Group(x.Key, (IEnumerable<ISelect2Option>)x.Select(c => c.CreateSelect2Option(TextSelector, IdSelector, OtherSelectors))));
                return new Select2Data(itens);
            }
            else
            {
                var itens = List.Select(c => c.CreateSelect2Option(TextSelector, IdSelector, OtherSelectors));
                return new Select2Data((IEnumerable<ISelect2Option>)itens);
            }
        }

        public static Select2Data CreateSelect2Data<OptionsType, T1, T2>(this PaginationFilter<T1, T2> Filter, Func<T2, string> TextSelector, Func<T2, string> IdSelector, Func<T2, string> GroupSelector = null, Action<T2, OptionsType> OtherSelectors = null)
            where OptionsType : ISelect2Option
            where T1 : class
        {
            var d = Filter.GetPage().CreateSelect2Data(TextSelector, IdSelector, OtherSelectors, GroupSelector);
            d.Pagination.More = Filter.PageCount > 1 && Filter.IsLastPage == false;
            return d;
        }

        public static Select2Data CreateSelect2Data<OptionsType, T1, T2>(this PaginationFilter<T1, T2> Filter, Func<T2, string> TextSelector, Func<T2, string> IdSelector, Action<T2, OptionsType> OtherSelectors)
            where OptionsType : ISelect2Option
            where T1 : class
        {
            return Filter.CreateSelect2Data(TextSelector, IdSelector, null, OtherSelectors);
        }

        public static Select2Data CreateSelect2Data<OptionsType, T1>(this PaginationFilter<T1, OptionsType> Filter)
            where OptionsType : ISelect2Option
            where T1 : class
        {
            return Filter.CreateSelect2Data<OptionsType, T1, OptionsType>((x) => x.Text, (x) => x.ID);
        }

        public static Select2Data CreateSelect2Data<OptionsType, T1>(this PaginationFilter<T1, OptionsType> Filter, Func<OptionsType, string> GroupBySelector)
            where OptionsType : ISelect2Option
            where T1 : class
        {
            return Filter.CreateSelect2Data<OptionsType, T1, OptionsType>((x) => x.Text, (x) => x.ID, GroupBySelector);
        }

        public static OptionType CreateSelect2Option<OptionType, T>(this T item, Func<T, string> TextSelector, Func<T, string> IdSelector, Action<T, OptionType> OtherSelectors = null) where OptionType : ISelect2Option
        {
            if (ReferenceEquals(typeof(T), typeof(OptionType)))
            {
                return item.ChangeType<OptionType>();
            }
            else
            {
                IdSelector = IdSelector ?? TextSelector;
                TextSelector = TextSelector ?? IdSelector;
                var Optionitem = Activator.CreateInstance<OptionType>();
                Optionitem.ID = IdSelector(item);
                Optionitem.Text = TextSelector(item);
                if (OtherSelectors != null)
                {
                    OtherSelectors(item, Optionitem);
                }

                return Optionitem;
            }
        }
    }

    public class Pagination
    {
        public bool More { get; set; } = false;
    }

    /// <summary>
    /// Classe utilizada para auxiliar nas respostas de requisições feitas por aAJAX através do select2.js
    /// </summary>
    public class Select2Data
    {
        public Select2Data()
        {
        }

        public Select2Data(IEnumerable<ISelect2Option> Options)
        {
            Results = Options ?? Array.Empty<ISelect2Option>();
        }

        public Select2Data(IEnumerable<ISelect2Option> Options, bool PaginationMore)
        {
            Results = Options ?? Array.Empty<ISelect2Option>();
            Pagination.More = PaginationMore;
        }

        public Select2Data(IEnumerable<Select2Group> Groups)
        {
            Results = Groups ?? Array.Empty<Select2Group>();
        }

        public Select2Data(IEnumerable<Select2Group> Groups, bool PaginationMore)
        {
            Results = Groups ?? Array.Empty<Select2Group>();
            Pagination.More = PaginationMore;
        }

        public Pagination Pagination { get; set; } = new Pagination();
        public IEnumerable<ISelect2Result> Results { get; set; } = Array.Empty<ISelect2Result>();
    }

    public class Select2Group : ISelect2Result
    {
        public Select2Group(string Text)
        {
            this.Text = Text;
        }

        public Select2Group(string Text, IEnumerable<ISelect2Option> Children)
        {
            this.Text = Text;
            this.Children = Children ?? Array.Empty<ISelect2Option>();
        }

        public Select2Group()
        {
        }

        public IEnumerable<ISelect2Option> Children { get; set; } = Array.Empty<ISelect2Option>();
        public string Text { get; set; }
    }

    public sealed class Select2Option : ISelect2Option
    {
        public Select2Option()
        {
        }

        public Select2Option(string Text, string Value)
        {
            ID = Value.IfBlank(Text);
            this.Text = Text;
        }

        public bool Disabled { get; set; }
        public string ID { get; set; }
        public bool Selected { get; set; }
        public string Text { get; set; }

        public override string ToString()
        {
            return $"<option value='{ID}'{Disabled.AsIf(" disabled")}{Selected.AsIf(" selected")}>{Text}</option>";
        }
    }
}